prisma/migrations/20230814233157_multiforward/migration.sql

```diff
@@ -0,0 +1,335 @@
+    END LOOP;
+
+    RETURN item;
+END;
+$$;
+
+DROP FUNCTION IF EXISTS update_poll(
+    sub TEXT, id INTEGER, title TEXT, text TEXT, boost INTEGER,
+    options TEXT[], fwd_user_id INTEGER);
+
+CREATE OR REPLACE FUNCTION update_poll(
+    sub TEXT, id INTEGER, title TEXT, text TEXT, boost INTEGER,
+    options TEXT[], forward JSON)
+RETURNS "Item"
+LANGUAGE plpgsql
+AS $$
+DECLARE
+    item "Item";
+    option TEXT;
+BEGIN
+    PERFORM ASSERT_SERIALIZED();
+
+    item := update_item(sub, id, title, null, text, boost, null, forward);
+
+    FOREACH option IN ARRAY options LOOP
+        INSERT INTO "PollOption" (created_at, updated_at, "itemId", "option") values (now_utc(), now_utc(), item.id, option);
+    END LOOP;
+
+    RETURN item;
+END;
+$$;
+
+-- Update item_act to support multi-way forward splits
+CREATE OR REPLACE FUNCTION item_act(item_id INTEGER, user_id INTEGER, act "ItemActType", act_sats INTEGER)
+RETURNS INTEGER
+LANGUAGE plpgsql
+AS $$
+DECLARE
+    user_msats BIGINT;
+    act_msats BIGINT;
+    fee_msats BIGINT;
+    item_act_id INTEGER;
+    referrer_id INTEGER;
+    fwd_entry record; -- for loop iterator variable to iterate across forward recipients
+    fwd_msats BIGINT; -- for loop variable calculating how many msats to give each forward recipient
+    total_fwd_msats BIGINT; -- accumulator to see how many msats have been forwarded for the act
+BEGIN
+    PERFORM ASSERT_SERIALIZED();
+
+    act_msats := act_sats * 1000;
+    SELECT msats, "referrerId" INTO user_msats, referrer_id FROM users WHERE id = user_id;
+    IF act_msats > user_msats THEN
+        RAISE EXCEPTION 'SN_INSUFFICIENT_FUNDS';
+    END IF;
+
+    -- deduct msats from actor
+    UPDATE users SET msats = msats - act_msats WHERE id = user_id;
+
+    IF act = 'VOTE' THEN
+        RAISE EXCEPTION 'SN_UNSUPPORTED';
+    END IF;
+
+    IF act = 'TIP' THEN
+        -- call to influence weightedVotes ... we need to do this before we record the acts because
+        -- the priors acts are taken into account
+        PERFORM weighted_votes_after_tip(item_id, user_id, act_sats);
+
+        -- take 10% and insert as FEE
+        fee_msats := CEIL(act_msats * 0.1);
+        act_msats := act_msats - fee_msats;
+
+        INSERT INTO "ItemAct" (msats, "itemId", "userId", act, created_at, updated_at)
+            VALUES (fee_msats, item_id, user_id, 'FEE', now_utc(), now_utc())
+            RETURNING id INTO item_act_id;
+
+        total_fwd_msats := 0; -- initialize accumulator to 0
+
+        -- add sats to actees' balance and stacked count
+        FOR fwd_entry IN SELECT "userId", "pct" FROM "ItemForward" WHERE "itemId" = item_id
+        LOOP
+            -- fwd_msats represents the sats for this forward recipient from this particular tip action
+            -- use FLOOR just to ensure we don't introduce more msats than we should, give the rest to OP
+            fwd_msats := FLOOR(act_msats * fwd_entry.pct / 100);
```

Comment:
I believe this floor is redundant as `act_msats * fwd_entry.pct` is coerced into a bigint and dividing an integer or a bigint by an integer yields an integer or bigint respectively.

It's nice to have the redundancy actually but worth noting I think.

Reply:
Good point, I am a little rusty with my typings in SQL so I wanted to ensure integer values here.